core/internal/api/credentials.go

```diff
@@ -55,3 +65,237 @@ func (c *apiKeyCredentialProvider) Apply(req *http.Request) error {
+		return err
+	}
+
+	parsedTime, err := time.Parse(expiresAtLayout, timeString)
+	if err != nil {
+		return err
+	}
+
+	*e = ExpiresAt(parsedTime)
+	return nil
+}
+
+func (e ExpiresAt) MarshalJSON() ([]byte, error) {
+	formattedTime := time.Time(e).Format(expiresAtLayout)
+	return json.Marshal(formattedTime)
+}
+
+type tokenInfo struct {
+	// The time at which the token will expire
+	ExpiresAt ExpiresAt `json:"expires_at"`
+
+	// The access token to use for authentication
+	AccessToken string `json:"access_token"`
+}
+
+func (c *tokenInfo) IsTokenExpiring() bool {
+	return time.Until(time.Time(c.ExpiresAt)) <= time.Minute*5
+}
+
+// CredentialsFile is used when serializing/deserializing JSON data from the
+// credentials file
+type CredentialsFile struct {
+	Credentials map[string]tokenInfo `json:"credentials"`
+}
+
+// Apply checks if the access token is expiring, and fetches a new one if so.
+// It then supplies the access token to the request via the Authorization header
+// as a Bearer token
+func (c *oauth2CredentialProvider) Apply(req *http.Request) error {
+	if c.token.IsTokenExpiring() {
+		if err := c.loadCredentials(); err != nil {
+			return err
+		}
+	}
+	req.Header.Set(
+		"Authorization",
+		"Bearer "+c.token.AccessToken,
+	)
+	return nil
+}
+
+// loadCredentials attempts to load an access token from the credentials file.
+// If the credentials file does not exist, it creates it.
+func (c *oauth2CredentialProvider) loadCredentials() error {
+	c.mu.Lock()
+	defer c.mu.Unlock()
+
+	_, err := os.Stat(c.credentialsFilePath)
+	if os.IsNotExist(err) {
+		err = c.writeCredentialsFile()
+	}
+	if err != nil {
+		return err
+	}
+
+	return c.loadCredentialsFromFile()
+}
+
+// loadCredentialsFromFile loads the access token from the credentials file. If
+// the access token does not exist for the given base url, or the token is
+// expiring, it fetches a new one from the server, and saves it to the credentials file
+func (c *oauth2CredentialProvider) loadCredentialsFromFile() error {
+	file, err := os.ReadFile(c.credentialsFilePath)
+	if err != nil {
+		return fmt.Errorf("failed to read credentials file: %v", err)
+	}
+
+	var credsFile CredentialsFile
+	if err := json.Unmarshal(file, &credsFile); err != nil {
+		return fmt.Errorf("failed to read credentials file: %v", err)
+	}
+
+	if credsFile.Credentials == nil {
+		credsFile.Credentials = make(map[string]tokenInfo)
+	}
+
+	creds, ok := credsFile.Credentials[c.baseURL]
+
+	if !ok || creds.IsTokenExpiring() {
+		newCreds, err := c.createAccessToken()
+		if err != nil {
+			return err
+		}
+		credsFile.Credentials[c.baseURL] = *newCreds
+		updatedFile, err := json.MarshalIndent(credsFile, "", "  ")
+		if err != nil {
+			return fmt.Errorf("failed to update credentials file: %v", err)
+		}
+		err = os.WriteFile(c.credentialsFilePath, updatedFile, 0600)
+		if err != nil {
+			return fmt.Errorf("failed to update credentials file: %v", err)
+		}
```

Comment:
Should a failure to save the credentials to a file prevent us from using the token? Maybe the filesystem is acting up, or there's a permissions problem with the directory.

Reply:
I think so. I don't want to get into a state where we are constantly creating new tokens. We also provide the flexibility to set the credentials file path, so if the default path doesn't work for them, they could adjust it

Reply:
What's the purpose of the file actually? Why not store the token in-memory?

If a user runs two different scripts in different terminal tabs, but both scripts use the same credentials file path, then the two `wandb-core` processes will race when writing the credentials file and could corrupt the data.

Besides the potential race, we do see file operations fail in strange ways quite often---no space left on disk, NFS issues, etc. File operations can also be very slow (again, network file systems!).

With all that in mind, is it worth using a file at all? And if it is, could we use an in-memory token anyway as it's faster & more reliable?

Reply:
We do store it in-memory and retrieve it from there in almost all cases. It is on startup when we would retrieve it from the file (if it exists). The purpose of the file is to have a back-up if the script crashes or just stops but the token hasn't expired yet, so we don't create new tokens each time. I've noticed aws boto3 and the gcloud sdk also do this. But I suppose there is an argument there that the file might be more convenient for interactive workloads than non-interactive workloads.

I am curious if we see the issues mentioned with the existing auth mode with the .netrc file?

Reply:
> We do store it in-memory and retrieve it from there in almost all cases. It is on startup when we would retrieve it from the file (if it exists).

`loadCredentials()` returns `loadCredentialsFromFile()`, so it fails if there are problems with the file. It also errors out if `writeCredentialsFile()` fails, but this should be best-effort.

> I am curious if we see the issues mentioned with the existing auth mode with the .netrc file?

I'm not sure! To be honest, I'm not completely certain we would see it in Sentry. For one, we only read .netrc at the start of the run, so there aren't many opportunities to fail. I'm also not sure if the code panics too quickly to be able to upload to Sentry.

Reply:
It's almost the same with the credentials file though. We read at the start of the run (or write if it doesn't exist). The other time we would read/write is when the token is close to expiring. @timoffex I think it's reasonable to make this best effort though. I will make that change.